File: python/ais_sdk/image_moderation_batch_jobs.py
# -*- coding:utf-8 -*-
import sys
import time
import json
import ais_sdk.ais as ais
import ais_sdk.utils as utils
import ais_sdk.signer as signer

#
# access moderation, image content of batch jobs，post data by token
#
_RETRY_TIMES = 3

def image_batch_jobs(token, urls, categories=['politics', 'terrorism', 'porn']):
    endpoint = utils.get_endpoint(ais.AisService.MODERATION_SERVICE)
    status, r = _image_batch_jobs(endpoint, token, urls, categories)

    if status != 200:
        return r

    submit_result = json.loads(r)
    job_id = submit_result['result'].get('job_id', '')
    # print "Process job id is :", job_id
    time.sleep(1.0)

    retry_times = 0
    try:
        while True:
            status, r = _get_result(endpoint, token, job_id)
            if status != 200:
                if retry_times < _RETRY_TIMES:
                    retry_times += 1
                    time.sleep(2.0)
                    continue
                else:
                    return r

            rec_result = json.loads(r)

            process_status = rec_result["result"].get('status')
            if process_status == 'failed':
                return r

            elif process_status == 'finish':
                return r

            #
            # process_status == 0 || process_status == 1
            #
            else:
                time.sleep(2.0)
                continue

    except Exception:
        return ''


#
# image content batch jobs, post the data
#
def _image_batch_jobs(endpoint, token, urls, categories=['politics', 'terrorism' ,'porn']):
    _url = 'https://%s/v1.0/moderation/image/batch/jobs' % endpoint

    _data = {
        "urls": urls,
        "categories": categories
    }

    status, resp = utils.request_token(_url, _data, token)
    if sys.version_info.major < 3:
        return status, resp
    else:
        return status, resp.decode('utf-8')


#
# access moderation, batch jobs, get the result
#
def _get_result(endpoint, token, job_id):
    _url_tmpl = 'https://%s/v1.0/moderation/image/batch?job_id=%s'
    _url = _url_tmpl % (endpoint, job_id)
    return utils.request_job_result_token(_url, token)



#
# access moderation, image content of batch jobs，post data by ak,sk
#
def image_batch_jobs_aksk(_ak, _sk, urls, categories=['politics', 'terrorism']):
    sig = signer.Signer()
    sig.AppKey = _ak
    sig.AppSecret = _sk

    endpoint = utils.get_endpoint(ais.AisService.MODERATION_SERVICE)

    status, r = _image_batch_jobs_aksk(endpoint, sig, urls, categories)

    if status != 200:
        return r

    submit_result = json.loads(r)
    job_id = submit_result['result'].get('job_id', '')
    # print "Process job id is :", job_id
    time.sleep(1.0)
    retry_times = 0
    try:
        while True:
            status, r = _get_result_aksk(endpoint, sig, job_id)
            if status != 200:
                if retry_times < _RETRY_TIMES:
                    retry_times += 1
                    time.sleep(2.0)
                    continue
                else:
                    return r

            rec_result = json.loads(r)

            process_status = rec_result["result"].get('status')
            if process_status == 'failed':
                return r

            elif process_status == 'finish':
                return r

            #
            # process_status == 0 || process_status == 1
            #
            else:
                time.sleep(2.0)
                continue

    except Exception:
        return ''


#
# image content of batch jobs, post the data
#
def _image_batch_jobs_aksk(endpoint, sig, urls, categories=['politics', 'terrorism']):
    _url = 'https://%s/v1.0/moderation/image/batch/jobs' % endpoint

    _data = {
        "urls": urls,
        "categories": categories
    }

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = endpoint
    kreq.uri = "/v1.0/moderation/image/batch/jobs"
    kreq.method = "POST"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.body = json.dumps(_data)

    status, resp = utils.request_aksk(sig, kreq, _url)
    if sys.version_info.major < 3:
        return status, resp
    else:
        return status, resp.decode('utf-8')


#
# access moderation, batch jobs, get the result
#
def _get_result_aksk(endpoint, sig, job_id):
    _url_tmpl = 'https://%s/v1.0/moderation/image/batch?job_id=%s'
    _url = _url_tmpl % (endpoint, job_id)

    kreq = signer.HttpRequest()
    kreq.scheme = "https"
    kreq.host = endpoint
    kreq.uri = "/v1.0/moderation/image/batch"
    kreq.method = "GET"
    kreq.headers = {"Content-Type": "application/json"}
    kreq.query = {'job_id': job_id}

    return utils.request_job_result_aksk(sig, kreq, _url)

File: javascript/ais_sdk/image_moderation_batch_jobs.js
var https = require("https");
var utils = require("./utils");
var signer = require("./signer");
var ais = require("./ais");

function batch_jobs(token, urls, categories, callback) {
    /**
     *  urls：图片的URL路径，目前支持对服务授权访问华为云上OBS的URL Array
     *  categories：非必选 检测场景 Array politics：是否涉及政治人物的检测。terrorism：是否包含暴恐元素的检测。porn：是否包
     *  含涉黄内容元素的检测。默认检测politics和terrorism
     * @type {string}
     */
    var requestData = {"urls": urls, "categories": categories};
    var requestBody = JSON.stringify(requestData);

    var endPoint = utils.getEndPoint(ais.MODERATION_SERVICE);

    // 构建请求信息
    var headers = {"Content-Type": "application/json", "X-Auth-Token": token, "Content-Length": requestBody.length};
    var options = utils.getHttpRequestEntityOptions(endPoint, "POST", ais.IMAGE_CONTENT_BATCH_JOBS, headers);

    var request = https.request(options, function (response) {

        response.on("data", function (chunk) {

            // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
            if (response.statusCode !== 200) {
                console.log('Http status code is: ' + response.statusCode);
                console.log(chunk.toString());
                return;
            }

            // 获取job_id
            var result = JSON.parse(chunk);
            job_id = result.result.job_id;
            console.log('Process job id is :' + job_id);
            var retryTimes = 0;

            // 根据job_id的结果,获取批量异步图像内容审核结果
            var results = "";
            get_result(endPoint, job_id, results, token, retryTimes, callback);
        })
    });

    request.on("error", function (err) {
        console.log(err.message);
    });

    request.write(requestBody);
    request.end();
}


function get_result(endPoint, job_id, resultsearch, token, retryTimes, callback) {
    // 构建请求参数和请求信息
    var requestData = {'job_id': job_id};
    var options = utils.getHttpRequestEntityForGet(endPoint, "GET", ais.IMAGE_CONTENT_BATCH_RESULT, {
        "Content-Type": "application/json",
        "X-Auth-Token": token,
    }, requestData);

    // 轮询请求图像内容审核接口，获取结果信息
    var reqsearch = https.request(options, function (res) {
        res.setEncoding('utf8');
        res.on('data', function (chunk) {
            resultsearch = JSON.parse(chunk);
            if (res.statusCode !== 200) {
                if(retryTimes < ais.RETRY_TIMES_MAX){
                    retryTimes++;
                    // 满足条件进行重试
                    setTimeout(function () {
                        get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback);
                    }, 2000);
                    console.log('The processing job request failed, retrying');
                }else {
                    console.log('Http status code is: ' + response.statusCode);
                    console.log(chunk.toString());
                }
                return;
            }

            // 如果处理失败，直接退出
            if (resultsearch !== "" && resultsearch.result.status === "failed") {
                console.log('The processing job request failed');
                callback(JSON.stringify(resultsearch));
                return;

            }

            // 任务处理成功
            if (resultsearch !== "" && resultsearch.result.status === "finish") {
                callback(JSON.stringify(resultsearch));
            }
            else {

                // 如果没有返回，等待一段时间，继续进行轮询。
                setTimeout(function () {
                    get_result(endPoint, job_id, resultsearch, token, retryTimes, callback);

                }, 2000);
            }
        });
    });

    reqsearch.on('error', function (e) {
        console.log('problem with request: ' + e.message);
    });

    reqsearch.write("");
    reqsearch.end();
}

function batch_jobs_aksk(_ak, _sk, urls, categories, callback) {
    // 配置ak，sk信息
    var sig = new signer.Signer();
    sig.AppKey = _ak;                     // 构建ak
    sig.AppSecret = _sk;                  // 构建sk

    var endPoint = utils.getEndPoint(ais.MODERATION_SERVICE);

    // 构建请求信息和请求参数信息
    var requestData = {"urls": urls, "categories": categories};
    var job_id = "";
    var req = new signer.HttpRequest();
    var header = {"Content-Type": "application/json"};
    var options = utils.getHttpRequestEntity(sig, req, endPoint, "POST", ais.IMAGE_CONTENT_BATCH_JOBS, "", header, requestData);

    var request = https.request(options, function (response) {

        response.on("data", function (chunk) {

            // 验证服务调用返回的状态是否成功，如果为200, 为成功, 否则失败。
            if (response.statusCode !== 200) {
                console.log('Http status code is: ' + response.statusCode);
                console.log(chunk.toString());
                return;
            }

            // 根据job_id的结果,获取批量异步图像内容审核结果
            var result = JSON.parse(chunk);
            job_id = result.result.job_id;
            console.log('Process job id is :' + job_id);

            var retryTimes = 0;
            var results = "";
            get_result_aksk(endPoint, sig, job_id, results,retryTimes, callback);
        })
    });

    request.on("error", function (err) {
        console.log(err.message);
    });

    request.write(req.body);
    request.end();
}


/**
 * 通过job_id 的信息，获取图像内容审核结果
 * @param endPoint ：服务域名
 * @param sign : 签名对象
 * @param job_id : 任务的id
 * @param resultsearch ：获取结果信息
 * @param retryTimes ：失败重试次数
 * @param callback ：回调函数
 */
function get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback) {

    // 构建请求信息和参数信息
    var request = new signer.HttpRequest();
    var options = utils.getHttpRequestEntity(sign, request, endPoint, "GET", ais.IMAGE_CONTENT_BATCH_RESULT, {'job_id': job_id}, {"Content-Type": "application/json"}, "");

    // 轮询图像内容审核接口，获取结果信息
    var reqsearch = https.request(options, function (response) {

        response.setEncoding('utf8');
        response.on('data', function (chunk) {
            resultsearch = JSON.parse(chunk);
            if (response.statusCode !== 200) {
                if(retryTimes < ais.RETRY_TIMES_MAX){
                    retryTimes++;
                    // 满足条件进行重试
                    setTimeout(function () {
                        get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback);
                    }, 2000);
                    console.log('The processing job request failed, retrying');
                }else {
                    console.log('Http status code is: ' + response.statusCode);
                    console.log(chunk.toString());
                }
                return;
            }

            // 如果处理失败，直接退出
            if (resultsearch !== "" && resultsearch.result.status === "failed") {
                console.log('The processing job request failed');
                callback(JSON.stringify(resultsearch));
                return;
            }

            // 任务处理成功
            if (resultsearch !== "" && resultsearch.result.status === "finish") {
                callback(JSON.stringify(resultsearch));
            }
            else {

                // 如果没有返回，等待一段时间，继续进行轮询。
                setTimeout(function () {
                    get_result_aksk(endPoint, sign, job_id, resultsearch, retryTimes, callback);
                }, 2000);
            }
        });
    });

    reqsearch.on('error', function (e) {
        console.log('problem with request: ' + e.message);

    });

    reqsearch.end();
}

module.exports = {
    batch_jobs,
    batch_jobs_aksk
};
